server/api_helper.go

```diff
@@ -0,0 +1,105 @@
+package main
+
+import (
+	"github.com/mattermost/mattermost-server/model"
+)
+
+// GetAllUsers returns all users
+func (p *Plugin) GetAllUsers() (map[string]*model.User, *model.AppError) {
+	allUsers := make(map[string]*model.User)
+	teamUsers, err := p.getTeamUsers()
+	if err != nil {
+		return nil, err
+	}
+	for _, users := range teamUsers {
+		for _, user := range users {
+			allUsers[user.Id] = user
+		}
+	}
+	return allUsers, nil
+}
+
+// GetAllChannels returns all channels
+func (p *Plugin) GetAllChannels() (map[string]*model.Channel, *model.AppError) {
+	allChannels := make(map[string]*model.Channel)
+	teamUsers, err := p.getTeamUsers()
+	if err != nil {
+		return nil, err
+	}
+	for team, users := range teamUsers {
+		for _, user := range users {
+			channels, err := p.API.GetChannelsForTeamForUser(team, user.Id, true)
+			if err != nil {
+				return nil, err
+			}
+			for _, channel := range channels {
+				allChannels[channel.Id] = channel
+			}
+		}
+	}
+	return allChannels, nil
+}
+
+// GetAllPublicChannelsForUser returns all public channels for user
+func (p *Plugin) GetAllPublicChannelsForUser(userID string) (map[string]*model.Channel, *model.AppError) {
+	allPublicChannels := make(map[string]*model.Channel)
+	teams, err := p.API.GetTeamsForUser(userID)
+
+	if err != nil {
+		return nil, err
+	}
+	for _, team := range teams {
+		perPage := 100
+		for page := 0; ; page++ {
+			channelsForTeam, err := p.API.GetPublicChannelsForTeam(team.Id, page, perPage)
+			if err != nil {
+				return nil, err
+			}
+			if len(channelsForTeam) == 0 {
+				break
+			}
+			for _, channel := range channelsForTeam {
+				allPublicChannels[channel.Id] = channel
+			}
+		}
+	}
+	return allPublicChannels, nil
+}
+
+// getTeamUsers returns slice of users for every team
+func (p *Plugin) getTeamUsers() (map[string][]*model.User, *model.AppError) {
+	teamUsers := make(map[string][]*model.User)
+	teams, err := p.API.GetTeams()
+	if err != nil {
+		return nil, err
+	}
+	for _, team := range teams {
+		page := 0
+		perPage := 100
+		for {
+			users, err := p.API.GetUsersInTeam(team.Id, page, perPage)
+			if err != nil {
+				return nil, err
+			}
+			if len(users) == 0 {
+				break
+			}
+			teamUsers[team.Id] = append(teamUsers[team.Id], users...)
+			page++
+		}
+	}
+	return teamUsers, nil
+}
+
+// RunOnSingleNode will run function f on only a single node in a HA environment
+func (p *Plugin) RunOnSingleNode(f func()) error {
+	ok, err := p.Helpers.KVCompareAndSetJSON("RunOnSingleNode", nil, 0)
+	if err != nil {
+		return err
+	}
+	if !ok {
+		f()
+	}
+	p.Helpers.KVCompareAndSetJSON("RunOnSingleNode", 0, nil) //TODO what to do if not saved?
```

Comment:
As a short term workaround to having automatic support for the expiration, what if instead of setting the value `0`, we set the current timestamp, and if we fail to acquire, check the timestamp and if it's greater than some threshold,  try to `KVCompareAndSetJSON` using that value instead?

Reply:
It's a distributed consensus problem, so there is no easy solution. This workaround will increase complexity but will have the same flaw - `KVCompareAndSetJSON` not working. And in addition we will be relying on the timestamp working right(synchronized) on every node. Probably better to use `KVCompareAndSetWithExpiryJSON`?

For this concrete plugin current implementation is not a big problem, because if `KVCompareAndSetJSON` will not delete the key on the first time cron job will try to do it again.
I'm more concerned about the `RunOnSingleNode` implementation as a standalone function.

Reply:
@iomodo, can you elaborate on this being a distributed consensus problem? `CompareAndSet` avoids the distributed issues by relying on an atomic write to the database. It is impossible for more than one node to race with another node using `CompareAndSet` and believe they both write a /new/ value.

I think we can rely on a reasonable degree of synchronization with respect to the timestamps between servers, but I note that the correctness of the atomicity doesn't rely on this, only the decision to forcibly expire the lock previously acquired. If the timestamps are off by more than a few seconds, posts from one server will appear out of sync with posts from another, and we have a larger problem at the application level.

`KVCompareAndSetWithExpiryJSON` doesn't help here, since it's just literally the above algorithm embedded into the server. It probably could (should?) have been a plugin helper instead.